### PR TITLE
Fix false-positive in `use_self`

### LINKED ISSUE #6818: use_self is triggered inside derived code
It looks like `use_self` is triggered inside derived macro-generated code. While derived macro authors may handle that on their part, it is still not feasible to expect that to happen overnight. Right now everyone wishing to use this lint inside their own code is affected.

Lint name: use_self

I tried this code:
```rust
#![deny(clippy::use_self)]

use serde::{Serialize, Deserialize};

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
struct Text {
    text: String,    
}

fn main() {
    let text = Text { text: String::from("bar") };
    println!("{:?}", text);
}
```

I expected to see this happen: clean cargo clippy run

Instead, this happened:
```
    Checking playground v0.0.1 (/playground)
error: unnecessary structure name repetition
 --> src/main.rs:6:8
  |
6 | struct Text {
  |        ^^^^ help: use the applicable keyword: `Self`
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![deny(clippy::use_self)]
  |         ^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#use_self

error: aborting due to previous error

error: could not compile `playground`

To learn more, run the command again with --verbose.
```

### Meta

- `cargo clippy -V`: clippy 0.1.52 (e37a13c 2021-02-28)
- `rustc -Vv`:
  ```
rustc 1.52.0-nightly (e37a13cc3 2021-02-28)
binary: rustc
commit-hash: e37a13cc3594004663738bd18d8100e6db9666cf
commit-date: 2021-02-28
host: x86_64-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 11.0.1
  ```

### PR DESCRIPTION
**Pull request by daxpedda**
Fixes  #6818. 

changelog: Fix false positives for use_self in macros.

### CODE REVIEW

File: tests/ui/use_self.rs [hunk 1]
```diff
@@ -454,3 +454,10 @@ mod nested_paths {
         }
     }
 }
+
+mod issue6818 {
+    #[derive(serde::Deserialize)]
```
Comment (daxpedda) [inline tests/ui/use_self.rs:459]: I was unable to generate a proper testcase for this, so I used this instead until somebody (or myself) can figure this out.

Comment (Manishearth) [reply tests/ui/use_self.rs:459]: Why is this not a proper testcase?

**Review by Manishearth — APPROVED** [review]
Approach lgtm

File: tests/ui/use_self.rs (hunk 1 above)
Comment (daxpedda) [reply tests/ui/use_self.rs:459]: I don't know if we should rely on `serde` always producing that specific issue, they might change something in their macro that doesn't trigger a false-positive anymore.

I would have preferred if we could reproduce the specific code that causes this. I gave it another stab today without success.

Comment (Manishearth) [reply tests/ui/use_self.rs:459]: I think this is fine for now.

**Pull request merged**